File: Data Structures & Algorithms/house-robber-ii/submission-9.cpp
#include <unordered_map>
#include <algorithm>
class Solution {
private:
std::unordered_map<int,int> memo;
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) return nums[0];
        std::vector<int> clone = nums;
        clone.pop_back();
        int res_1 = dfs(nums,1);
        memo.clear();
        int res_2 = dfs(clone,0);
        return std::max(res_1,res_2);
    }
    int dfs(const std::vector<int> &nums, int idx){
        if (memo.count(idx)==1){
            return memo[idx];
        }
        if (idx >= nums.size()){
            return 0;
        }
        

        memo[idx] = std::max(dfs(nums,idx+1),nums[idx]+dfs(nums,idx+2));
        return memo[idx];
    }
};

File: Data Structures & Algorithms/decode-ways/submission-9.py
class Solution:
    def numDecodings(self, s: str) -> int:
        memo = {}
        if s == "" or s[0] == "0":
            return 0

        def dfs(s,i):
            if len(s) == 0:
                return 1
            if s[0] == "0":
                return 0
            if len(s) == 1:
                return 1
            if i in memo:
                return memo[i]

            if (s[0] == "1" and s[1]!="0") or (s[0] == "2" and (s[1]!= "0" and s[1]!= "7" and s[1]!= "8" and s[1]!= "9")): 
                memo[i] = dfs(s[1:],i+1) + dfs(s[2:],i+2)
            elif (s[0] == "1" and s[1]=="0") or (s[0]=="2" and s[1]=="0"):
                memo[i] = dfs(s[2:],i+2)
            else:
                memo[i] = dfs(s[1:],i+1)

            return memo[i]
        return dfs(s,0)

            
